### ISSUE TITLE
xsd for the checkstyle configuration xml file

### ISSUE THREAD

**Issue by debashisdeb** [checkstyle/checkstyle#7517]
**Use Case :**

I'm planning to introduce custom checks for multiple users, in which each project owner can define their own set of checks. I also have a set of default set of checks that needs to run for every project. During the run time of Checkstyle, I want to merge the project specific checks and the default checks. Having a XSD can help me validate if the merged XML configuration file is formed correctly or not.
Since some of the checks go inside of the Checker module and many checks go inside the TreeWalker module, we need to correctly place the checks, so that the run doesn't fail at all after creating a wrong merged XML.

**Question :**
Is there any XSD available for validating ? Or may be an API within Checkstyle which can do the merge ?
I've seen the DTD that is present, but that doesn't help in completely validating the entire configuration file.

**Comment by debashisdeb** [comment]
@romani  any ideas around this ?

**Comment by romani** [comment]
Yes, idea to migrate to xsd was around for very long time, but nobody actually willing to start this process.
here is request for XSD for reports - https://github.com/checkstyle/checkstyle/issues/5166

> Is there any XSD available for validating ?

NO.
DTD do validation of xml files for structure, it is old fashion approach, but it works.
So nothing is blocking you right now to generate config file and simply use latest version of DTD (they are stored as resources in checkstyle jar) to validate it (you can see our code how we validate file before parsing it).  

> Or may be an API within Checkstyle which can do the merge ?

No, merge of configs is very complicated concept, we do not know how to make it for now.
#6942 , #2873

> , but that doesn't help in completely validating the entire configuration file.

please explain why, we do such DTD by our code and by maven plugs to validate xml files.


------------------

Here is example on how users hacking config file to have it more modular (specific config for specific project) - #3605  

---- 

if you end up in creation of xsd files (based on our DTD), please contribute them to checkstyle.

**Comment by romani** [comment]
issue is not complete, if no activity happened till end of February, issue will be closed.

**Comment by xenoterracide** [comment]
huh, doesn't look closed, I just came wondering if this existed, would be good for autocompletion in IDE's

**Comment by romani** [comment]
@xenoterracide , which IDE you use ?
Is dtd not enough?

**Comment by xenoterracide** [comment]
Intellij, I guess it's suggesting tags, but not properties.

**Comment by romani** [comment]
> values of properties

good point to know ... yes I agree it would be cool.
But this will force us to have schema to be version specific and bound to checkstyle version, as new values appear all the time, new Checks appear each version. It will hard for us to keep this XSD uptoday manually, so we need generator of it.
 
But current DTD is stable for very long time.

All DTDs are downloadable by URL, as we have small amount of them it is manageable, with XSDs for each checkstyle version, amount will be significant, and users need update version of XSD each time they update checkstyle.

I am not sure final result cost all efforts to make it happen.
I keep issue open for ideas and discussions....

**Comment by xenoterracide** [comment]
xsd's are downloadable and you can point them at something like a latest url

**Comment by romani** [comment]
We releasing very frequently(each month), most users use very old versions.

**Comment by Potherca** [comment]
folder with test expected reports: https://github.com/checkstyle/checkstyle/tree/master/src/test/resources/com/puppycrawl/tools/checkstyle/xmllogger 

Basically something like this, right?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema">

  <!-- Root element -->
  <xs:element name="checkstyle" type="checkstyleType"/>

  <!-- checkstyleType (root type) -->
  <xs:complexType name="checkstyleType">
    <xs:sequence>
      <xs:element maxOccurs="unbounded" minOccurs="0" name="file" type="fileType"/>
    </xs:sequence>
    <xs:attribute name="version" type="xs:string" use="optional"/>
  </xs:complexType>

  <!-- fileType -->
  <xs:complexType name="fileType">
    <xs:sequence>
      <xs:element maxOccurs="unbounded" minOccurs="0" name="error" type="errorType"/>
    </xs:sequence>
    <xs:attribute name="name" type="xs:string" use="required"/>
  </xs:complexType>

  <!-- errorType -->
  <xs:complexType name="errorType">
    <xs:attribute name="column" type="xs:positiveInteger" use="optional"/>
    <xs:attribute name="line" type="xs:positiveInteger" use="optional"/>
    <xs:attribute name="message" type="xs:string" use="optional"/>
    <xs:attribute name="severity" type="severityType" use="optional"/>
    <xs:attribute name="source" type="xs:string" use="optional"/>
  </xs:complexType>

  <!-- severityType enum -->
  <xs:simpleType name="severityType">
    <xs:restriction base="xs:string">
      <xs:enumeration value="error"/>
      <xs:enumeration value="info"/>
      <xs:enumeration value="warning"/>
    </xs:restriction>
  </xs:simpleType>

</xs:schema>
```

**Comment by romani** [comment]
config file DTD is converted by https://code.google.com/archive/p/jing-trang/downloads
`$ java -jar trang.jar -I dtd -O xsd configuration_1_3.dtd  configuration_1_3.xsd`
plus some minor changes

**Comment by romani** [comment]
xsd at https://github.com/checkstyle/checkstyle/blob/master/config/configuration-1-3.xsd

**Comment by xenoterracide** [comment]
it's been a long time, what's the expected import for this? can we get some documentation on checkstyle.org? I looked here https://checkstyle.org/config.html

**Comment by romani** [comment]
not sure what you mean by "expected import".
page content is at https://github.com/checkstyle/checkstyle/blob/master/src/site/xdoc/config.xml please send PR

**Comment by Potherca** [comment]
I think that what @xenoterracide means with expected import, is:

```xml
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="path/to/checkstyle-report-1.0.0.xsd"
```

Where `path/to/` is either local or `https://...`.

**Comment by romani** [comment]
@Potherca, please create new issue to deploy this xsd to website
